xds/src/main/java/io/grpc/xds/EdsLoadBalancer.java

```diff
@@ -0,0 +1,715 @@
+
+      for (Locality oldLocality : localityMap.keySet()) {
+        if (!newLocalities.contains(oldLocality)) {
+          deactivate(oldLocality);
+        }
+      }
+    }
+
+    private void updateDropPercentage(List<DropOverload> dropOverloads) {
+      this.dropOverloads = checkNotNull(dropOverloads, "dropOverloads");
+    }
+
+    private void deactivate(final Locality locality) {
+      if (!localityMap.containsKey(locality) || localityMap.get(locality).isDeactivated()) {
+        return;
+      }
+
+      final LocalityLbInfo localityLbInfo = localityMap.get(locality);
+      class DeletionTask implements Runnable {
+
+        @Override
+        public void run() {
+          localityLbInfo.shutdown();
+          localityMap.remove(locality);
+        }
+
+        @Override
+        public String toString() {
+          return "DeletionTask: locality=" + locality;
+        }
+      }
+
+      localityLbInfo.delayedDeletionTimer = helper.getSynchronizationContext().schedule(
+          new DeletionTask(), DELAYED_DELETION_TIMEOUT_MINUTES,
+          TimeUnit.MINUTES, helper.getScheduledExecutorService());
+    }
+
+    // TODO(zdapeng): Decide how to work with LRS client.
+    LoadStatsStore getLoadStatsStore() {
+      return loadStatsStore;
+    }
+
+    void updateOobMetricsReportInterval(long reportIntervalNano) {
+      metricsReportIntervalNano = reportIntervalNano;
+      for (LocalityLbInfo lbInfo : localityMap.values()) {
+        lbInfo.childHelper.updateMetricsReportInterval(reportIntervalNano);
+      }
+    }
+
+    @Nullable
+    private static ConnectivityState aggregateState(
+        @Nullable ConnectivityState overallState, ConnectivityState childState) {
+      if (overallState == null) {
+        return childState;
+      }
+      if (overallState == READY || childState == READY) {
+        return READY;
+      }
+      if (overallState == CONNECTING || childState == CONNECTING) {
+        return CONNECTING;
+      }
+      if (overallState == IDLE || childState == IDLE) {
+        return IDLE;
+      }
+      return overallState;
+    }
+
+    private void updatePicker(
+        @Nullable ConnectivityState state,  List<WeightedChildPicker> childPickers) {
+      childPickers = Collections.unmodifiableList(childPickers);
+      SubchannelPicker picker;
+      if (childPickers.isEmpty()) {
+        if (state == TRANSIENT_FAILURE) {
+          picker = new ErrorPicker(Status.UNAVAILABLE); // TODO: more details in status
+        } else {
+          picker = XdsSubchannelPickers.BUFFER_PICKER;
+        }
+      } else {
+        picker = pickerFactory.picker(childPickers);
+      }
+
+      if (!dropOverloads.isEmpty()) {
+        picker = new DroppablePicker(dropOverloads, picker, random, loadStatsStore);
+      }
+
+      if (state != null) {
+        helper.updateBalancingState(state, picker);
+      }
+    }
+
+    @VisibleForTesting // Introduced for testing only.
```

Comment:
nit: does this comment really makes a lot of sense? I mean it literally duplicates the meaning of annotation itself

Reply:
"Introduced for testing only" has a little more information than `@VisibleForTesting`. `@VisibleForTesting` does not imply "Introduced for testing only".

Reply:
I agree, that annotation name itself doesn't imply that, but what do you think about javadoc? It says - "Annotates a program element that exists, or is more widely visible than otherwise necessary, **only for use in test code**"